python/ray/workflow/workflow_state.py

```diff
@@ -134,7 +136,12 @@ class WorkflowExecutionState:
     # already complete; if the task that is the root of all continuation completes,
     # then all its continuations would complete.
     done_tasks: Set[TaskID] = field(default_factory=set)
-
+    # The set of staged tasks. The dependencies (upstream inputs etc.) of these
+    # tasks have been constructed and they are ready to be submitted later.
+    # Once a task is submitted, it is moved out of this set.
+    # This field is checked and updated when reconstruct dependencies of failed
+    # tasks to eliminate concurrent issues like reconstructing a task multiple times.
+    staged_tasks: Set[TaskID] = field(default_factory=set)
```

Comment:
There are so many status for a task. Could you add some comment at the beginning of this `scheduling` session about how the status changes? An state machine will be good, for example:

`running_frontier --- when task finished --> staged_tasks` ??